spec/datadog/tracing/contrib/pg/patcher_spec.rb

```diff
@@ -0,0 +1,788 @@
+# typed: ignore
+
+require 'datadog/tracing/contrib/integration_examples'
+require 'datadog/tracing/contrib/support/spec_helper'
+require 'datadog/tracing/contrib/analytics_examples'
+
+require 'ddtrace'
+require 'pg'
+
+RSpec.describe 'PG::Connection patcher' do
+  let(:service_name) { 'pg' }
+  let(:configuration_options) { { service_name: service_name } }
+
+  let(:conn) do
+    PG::Connection.new(
+      host: host,
+      port: port,
+      dbname: dbname,
+      user: user,
+      password: password
+    )
+  end
+
+  let(:host) { ENV.fetch('TEST_POSTGRES_HOST') { '127.0.0.1' } }
+  let(:port) { ENV.fetch('TEST_POSTGRES_PORT') { '5432' } }
+  let(:dbname) { ENV.fetch('TEST_POSTGRES_DB') { 'postgres' } }
+  let(:user) { ENV.fetch('TEST_POSTGRES_USER') { 'postgres' } }
+  let(:password) { ENV.fetch('TEST_POSTGRES_PASSWORD') { 'postgres' } }
+
+  before do
+    Datadog.configure do |c|
+      c.tracing.instrument :pg, configuration_options
+    end
+  end
+
+  around do |example|
+    # Reset before and after each example; don't allow global state to linger.
+    Datadog.registry[:pg].reset_configuration!
+    example.run
+    Datadog.registry[:pg].reset_configuration!
+  end
+
+  after do
+    conn.close
+  end
+
+  describe 'tracing' do
+    describe '#exec' do
+      context 'when the tracer is disabled' do
+        before { tracer.enabled = false }
+
+        it 'does not write spans' do
+          conn.exec('SELECT 1;')
+          expect(spans).to be_empty
+        end
+      end
+
+      context 'when the tracer is configured directly' do
+        let(:service_override) { 'pg-override' }
+
+        before do
+          Datadog.configure_onto(conn, service_name: service_override)
+          conn.exec('SELECT 1;')
+        end
+
+        it 'produces a trace with service override' do
+          expect(spans.count).to eq(1)
+          expect(span.service).to eq(service_override)
+          expect(span.get_tag(Datadog::Tracing::Metadata::Ext::TAG_PEER_SERVICE)).to eq(service_override)
```

Comment:
I noticed you were testing `conn.exec('SELECT 1;')` are in this section. The way we organize the code under test is by declaring it as a `subject`.

In this case, we can declare `subject(:exec) { conn.exec('SELECT 1;') }` for the `exec` subject.
Now we can simply call `exec` or `subject` to execute the code under test.

I'd also move `exec` calls from before blocks and into the test execution itself. This allows us to set up any `before` blocks more freely, without having to worry about which `before` block is executed in which order.

I recommend the same to be done for all the tests in this file.

You can read more about `subject` in RSpec here: https://relishapp.com/rspec/rspec-core/docs/subject/explicit-subject

```suggestion
    describe '#exec' do
      subject(:exec) { conn.exec('SELECT 1;') }
      
      context 'when the tracer is disabled' do
        before { tracer.enabled = false }

        it 'does not write spans' do
          exec
          expect(spans).to be_empty
        end
      end

      context 'when the tracer is configured directly' do
        let(:service_override) { 'pg-override' }

        before do
          Datadog.configure_onto(conn, service_name: service_override)
        end

        it 'produces a trace with service override' do
          exec
          expect(spans.count).to eq(1)
          expect(span.service).to eq(service_override)
          expect(span.get_tag(Datadog::Tracing::Metadata::Ext::TAG_PEER_SERVICE)).to eq(service_override)
```

Reply:
changed the tests to use subject, but not sure if I am using it right for the shared examples? e.g.:
https://github.com/DataDog/dd-trace-rb/blob/995421dd64977c65990e246b55711353d2694ab2/spec/datadog/tracing/contrib/pg/patcher_spec.rb#L103-L116

Reply:
That looks great! 👍